### DISCUSSION TITLE
Workflow shortcuts?

### DISCUSSION THREAD

**Discussion by dabrahams** [magit/magit#4839] [Ideas]
There are sequences of commands I find myself running all the time, and I wonder if these kinds of things should somehow be bundled into Magit features.  For example,

I just finished work on a branch.  It was merged.  I'm still on that branch.  Now I want to:
1. fetch from origin
2. reset main to origin/main
3. switch to main
4. spinoff a new branch for my next bit of work

Every time I go through this I think, “this is the sort of thing that Magit usually makes less tedious.”  Would something like that be in scope?

Another: I want to submit a PR fixing a particular thing I found in `main`.  
1. If I have any outstanding changes, I want to make a WIP commit with any outstanding changes
2. Do the same process as above

Another: I want to submit a PR with the work I've done:
1. Push the branch to origin (or my fork if I've set that up as an upstream default—I don't think such a setting exists?)  without prompting me
2. Open the URL github dumps to stdout for creating the PR, or if forge is set up, create the PR that way.

Another: I've just made a one-commit change on main.  I want to spin off a branch with a name derived from the first line of the commit message, push it to the default upstream, and create a PR.

Another: I've made that change but haven't committed it yet.  I want it to prompt me for the commit message and do all of the above (same command)

Would these kinds of things be in scope for Magit?

**Comment by dabrahams** [comment]
@tarsius bump.  There are maintainer-directed questions in here.  I have some memory of getting an answer to these at one point, but I must be mistaken, because I don't see an answer here.  If I'm not mistaken, sorry for the noise.  Thanks

**Comment by tarsius** [comment]
> I just finished work on a branch. It was merged. I'm still on that branch.

Try the "merge into" variant. That checks out the upstream branch and then merges and *deletes* the feature branch.

> Now I want to:
> 1. fetch from origin
> 2. reset main to origin/main
> 3. switch to main
> 4. spinoff a new branch for my next bit of work

~~The above took care of 1 through 3.~~ Hm, well not really -- it still helps I think.

It seems okay that you have to do 4 explicitly.

**Reply by dabrahams** [reply]
> it still helps I think.

I'll explore it. From your description, it's not very obvious to me how it helps.

So I'd really like answers to the questions I asked, which is not about how to do these things in magit, but to know if you consider these _kinds_ of things in scope. Another way of asking is, would patches for them be welcomed?

If not, I suppose Magit has some kind of extensibility (at least I've seen extension packages out there like `magit-imerge`).  Is there a guide to writing magit extensions somewhere?

**Reply by tarsius** [reply]
> I'll explore it. From your description, it's not very obvious to me how it helps.

When writing that description, and now as well, I did not fully understand your workflow description. I had to do some guessing and so it is not surprising that you in turn have difficulties understand my proposed/assumed alternative approach, after all it is rooted in some assumptions of mine.

> So I'd really like answers to the questions I asked, which is not about how to do these things in magit, but to know if you consider these kinds of things in scope. Another way of asking is, would patches for them be welcomed?

I recently closed another proposal/question by you, with the justification, that it was "out of scope", to which you responded by explaining that that terminology does not apply. So I am wary of saying that this here is "out of scope". (#5143, I've just added another reply there.)

But I will give you an honest answer.

More often than not, when you propose some feature, I have difficulties understanding your description and/or seeing why one would want that.

When I do (belief to) understand both, then I usually think the proposal takes things too far, tailoring for one legitimate but very particular workflow. So usually I do push back a bit, explaining that those things are already possible, but require more individual steps, than you had hoped, or point you to different possible workflows.

What makes this challenging for me, is that you don't usually take a "no" with just a brief/incomplete/faulty justification for an answer. You usually want to talk it out, and I feel, that this means, that over the years, I had to given your proposals a disproportionate amount of attention. As a result, usually one or two proposals by you remain open, despite me being fairly certain that I won't implement them as requested.

----

Your proposals usually seem to boil down to "I would like to do A and B (I always to B after A), but I have to perform these steps separately. Could you add a command that does A+B in one go?".

A major appeal of Magit is that it provides many such A+B commands, so it is entirely expected that many users, when finding that for some workflow that is not the case, ask for a new command that covers that use-case. So, generally speaking, these "workflow shortcuts" are completely "in scope".

But they do have to be considered *individually* nonetheless. I do have to draw the line somewhere, and I do not draw it at "this makes absolutely no sense whatsoever". This necessarily is a judgement call. Made by me. At a given time (I might change my mind later). It may involve saying "no", which can be disappointing for the other party. Unfortunately for you, your proposals often, but not always, do not make the cut.

Beside making one particular user happy, I also have to consider opportunity cost, increased complexity and thus maintenance cost, likelihood of follow up requests for further improvements, cost of saying "no", emotional cost of accusation of feature creep from trolls, my own well being, my own expectations from the software, etc.

-----

In this case you seem to want some "workflow shortcut", that does the following two things.

1. Wrap up working on feature A, because it has been completed. (Cleanup for A.)
2. Start working on the next thing, feature B. (Setup for B.)

I understand that this is something that happens a lot. It happens to me a lot. It happens to everyone a lot. Despite that I do not think the work involved in these two steps should be combined into one step/command. There is *value* in winding down task A and then being *done*. Even if is just for a few seconds, until one has moved on to setting things up to tackle task B.

There is value in keeping these two steps *separate*, even if it is only spiritual. There is also time to be gained by *combining* these two steps, but I think that gain is very small. So small, that even though I have to pay it myself all the time, I do not see value in optimizing it away.

Despite that, I can of course also understand the urge in others to have this very common sequence of steps streamline. I do not think it is wrong of you to want that. I don't think it would be a mistake for you to implement that in your own package/configuration. But it is something that I do not want to add to Magit.

**Reply by dabrahams** [reply]
Thanks for taking the time to write this very detailed response.  I'm sorry that my issues have been frustrating for you.  Based on this

> usually one or two proposals by you remain open, despite me being fairly certain that I won't implement them as requested.

I think you should just close them.  Having things marked open for long periods with no action (and especially, little chance of action ever) serves nobody.

No need to explain the factors that make it necessary to reject patches.  That's why I added

> I suppose Magit has some kind of extensibility (at least I've seen extension packages out there like magit-imerge). Is there a guide to writing magit extensions somewhere?

To which I never got an answer.  All this “talking out” is just about trying to chart a path for myself with these wishlist items.  If the Jonas doesn't want to do it, would I be able to implement it? Is it worth the effort? Is it worth the extra effort to make it fit in with magit and present it as a PR?  Those kinds of things.  If I don't get answers, I keep asking.

Despite  that I'm almost completely retired from programming as of February (so don't bother answering more questions—I'm not going to do more programming just to make programming easier!) Magit essential and really the one thing that keeps me stuck to Emacs when I _do_ program.  Great piece of work.

**Reply by tarsius** [reply]
> [...] No need to explain the factors that make it necessary to reject patches.

Thanks!

> I suppose Magit has some kind of extensibility (at least I've seen extension packages out there like magit-imerge). Is there a guide to writing magit extensions somewhere?

There's no such guide, I'm afraid. There are some difficult areas to tackle, if you want to extend Magit, and some of those are documented separately. Most prominently, if you want to implement menus like those of Magit, read the Transient manual (but I would recommend to also look at uses of that in Magit and maybe other packages). `magit-section.el` also has its own manual, but when it comes to section functionality, you probably cannot get around also looking into how magit-proper is using it. There is no separate manual for "how magit calls git", but it's covered in https://docs.magit.vc/magit/Calling-Git.html. Everything one level up might be of interest too (https://docs.magit.vc/magit/Plumbing.html).

One thing that will be of interest to you is "chaining magit commands and/or git invocations". To chain asynchronous git invocations use `(set-process-sentinel magit-this-process (lambda (process event) ...)`, or similar. In some cases you might want to call `magit-process-sentinel` in there, in others just `magit-refresh`, or neither. There are plenty of examples in Magit. (I am not particularly happy with this interface, but it has worked well enough for a long time.)

Aside from the areas mentioned above and some particularly hairy parts such as diffing and logging, large parts of Magit is just "doing things thoughtfully". I hope that a lot of that code isn't all that complicated and can be used for inspiration.

Generally my advice is to look for existing code, that does something similar, and then adapt it to your needs.

**Reply by dabrahams** [reply]
Thanks; as noted I'm not going to try to move forward with any of these things, but you might want to put this in the README or Magit manual for other people.
